### PR TITLE
Fix SuperPMI collect with getIntConfigValue

### LINKED ISSUE #53440: superpmi collection crash
Crash doing superpmi PMI collection using:

```
py c:\gh\runtime\src\coreclr\scripts\superpmi.py collect --pmi -assemblies c:\gh\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root -jit_name c:\gh\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\clrjit.dll -output_mch_path c:\bugs\spmicollect10\libraries.pmi.windows.x64.checked.mch -pmi_location c:\gh\jitutils\bin\pmi.dll --use_zapdisable
```

stack:
```
 	KernelBase.dll!RaiseFailFastException()	Unknown
 	coreclr.dll!_DbgBreakCheck(const char * szFile, int iLine, const char * szExpr, int fConstrained) Line 427	C++
 	coreclr.dll!_DbgBreakCheckNoThrow(const char * szFile, int iLine, const char * szExpr, int fConstrained) Line 534	C++
 	coreclr.dll!DbgAssertDialog(const char * szFile, int iLine, const char * szExpr) Line 695	C++
 	coreclr.dll!EEFilterException(_EXCEPTION_POINTERS * pExceptionPointers, void * unused) Line 10614	C++
 	[Inline Frame] coreclr.dll!RunWithErrorTrapFilter(_EXCEPTION_POINTERS * exceptionPointers, void *) Line 10907	C++
 	coreclr.dll!`CEEInfo::runWithErrorTrap'::`1'::filt$0() Line 10937	C++
 	coreclr.dll!__C_specific_handler(_EXCEPTION_RECORD * ExceptionRecord, void * EstablisherFrame, _CONTEXT * ContextRecord, _DISPATCHER_CONTEXT * DispatcherContext) Line 376	C++
 	ntdll.dll!RtlpExecuteHandlerForException()	Unknown
 	ntdll.dll!RtlDispatchException()	Unknown
 	ntdll.dll!KiUserExceptionDispatch()	Unknown
 	superpmi-shim-collector.dll!CompileResult::AddCall(const char * name) Line 92	C++
 	superpmi-shim-collector.dll!JitHost::getIntConfigValue(const wchar_t * key, int defaultValue) Line 60	C++
 	clrjit.dll!Compiler::compInit(ArenaAllocator * pAlloc, CORINFO_METHOD_STRUCT_ * methodHnd, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, InlineInfo * inlineInfo) Line 1862	C++
 	[Inline Frame] clrjit.dll!jitNativeCode::__l8::__Body::Run::__l6::__Body::Run(jitNativeCode::__l8::__Body::Run::__l5::__JITParam *) Line 7090	C++
 	clrjit.dll!`jitNativeCode'::`8'::__Body::Run(jitNativeCode::__l2::__JITParam * __JITpParam) Line 7101	C++
 	clrjit.dll!jitNativeCode(CORINFO_METHOD_STRUCT_ * methodHnd, CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags, void * inlineInfoPtr) Line 7125	C++
 	clrjit.dll!Compiler::fgInvokeInlineeCompiler::__l2::<lambda>(Compiler::fgInvokeInlineeCompiler::__l2::Param * pParam) Line 979	C++
 	coreclr.dll!`CEEInfo::runWithErrorTrap'::`6'::__Body::Run(RunWithErrorTrapFilterParam * pTrapParam) Line 10935	C++
 	coreclr.dll!CEEInfo::runWithErrorTrap(void(*)(void *) function, void * param) Line 10937	C++
 	[Inline Frame] clrjit.dll!Compiler::eeRunWithErrorTrap(void(*)(Compiler::fgInvokeInlineeCompiler::__l2::Param *)) Line 7847	C++
 	clrjit.dll!Compiler::fgInvokeInlineeCompiler(GenTreeCall * call, InlineResult * inlineResult) Line 993	C++
 	clrjit.dll!Compiler::fgMorphCallInlineHelper(GenTreeCall * call, InlineResult * result) Line 6654	C++
 	clrjit.dll!Compiler::fgMorphCallInline(GenTreeCall * call, InlineResult * inlineResult) Line 6537	C++
 	clrjit.dll!Compiler::fgInline() Line 158	C++
 	clrjit.dll!Phase::Run() Line 62	C++
 	[Inline Frame] clrjit.dll!DoPhase(Compiler *) Line 136	C++
 	clrjit.dll!Compiler::compCompile(void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags) Line 4664	C++
 	clrjit.dll!Compiler::compCompileHelper(CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags) Line 6455	C++
 	[Inline Frame] clrjit.dll!Compiler::compCompile::__l73::__Body::Run(Compiler::compCompile::__l2::__JITParam *) Line 5748	C++
>	clrjit.dll!Compiler::compCompile(CORINFO_MODULE_STRUCT_ * classPtr, void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags) Line 5752	C++
 	[Inline Frame] clrjit.dll!jitNativeCode::__l8::__Body::Run::__l6::__Body::Run(jitNativeCode::__l8::__Body::Run::__l5::__JITParam *) Line 7098	C++
 	clrjit.dll!`jitNativeCode'::`8'::__Body::Run(jitNativeCode::__l2::__JITParam * __JITpParam) Line 7101	C++
 	clrjit.dll!jitNativeCode(CORINFO_METHOD_STRUCT_ * methodHnd, CORINFO_MODULE_STRUCT_ * classPtr, ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, void * * methodCodePtr, unsigned int * methodCodeSize, JitFlags * compileFlags, void * inlineInfoPtr) Line 7125	C++
 	clrjit.dll!CILJit::compileMethod(ICorJitInfo * compHnd, CORINFO_METHOD_INFO * methodInfo, unsigned int flags, unsigned char * * entryAddress, unsigned int * nativeSizeOfCode) Line 279	C++
 	superpmi-shim-collector.dll!interceptor_ICJC::compileMethod(ICorJitInfo * comp, CORINFO_METHOD_INFO * info, unsigned int flags, unsigned char * * nativeEntry, unsigned int * nativeSizeOfCode) Line 62	C++
 	coreclr.dll!invokeCompileMethodHelper(EEJitManager * jitMgr, CEEInfo * comp, CORINFO_METHOD_INFO * info, CORJIT_FLAGS jitFlags, unsigned char * * nativeEntry, unsigned int * nativeSizeOfCode) Line 12659	C++
 	coreclr.dll!invokeCompileMethod(EEJitManager * jitMgr, CEEInfo * comp, CORINFO_METHOD_INFO * info, CORJIT_FLAGS jitFlags, unsigned char * * nativeEntry, unsigned int * nativeSizeOfCode) Line 12724	C++
 	coreclr.dll!`CallCompileMethodWithSEHWrapper'::`10'::__Body::Run(CallCompileMethodWithSEHWrapper::__l2::Param * pParam) Line 12778	C++
 	coreclr.dll!CallCompileMethodWithSEHWrapper(EEJitManager * jitMgr, CEEInfo * comp, CORINFO_METHOD_INFO * info, CORJIT_FLAGS flags, unsigned char * * nativeEntry, unsigned int * nativeSizeOfCode, NativeCodeVersion nativeCodeVersion) Line 12785	C++
 	coreclr.dll!UnsafeJitFunction(PrepareCodeConfig * config, COR_ILMETHOD_DECODER * ILHeader, CORJIT_FLAGS flags, unsigned long * pSizeOfCode) Line 13314	C++
 	coreclr.dll!MethodDesc::JitCompileCodeLocked(PrepareCodeConfig * pConfig, ListLockEntryBase<NativeCodeVersion> * pEntry, unsigned long * pSizeOfCode, CORJIT_FLAGS * pFlags) Line 1023	C++
 	coreclr.dll!MethodDesc::JitCompileCodeLockedEventWrapper(PrepareCodeConfig * pConfig, ListLockEntryBase<NativeCodeVersion> * pEntry) Line 892	C++
 	coreclr.dll!MethodDesc::JitCompileCode(PrepareCodeConfig * pConfig) Line 832	C++
 	coreclr.dll!MethodDesc::PrepareILBasedCode(PrepareCodeConfig * pConfig) Line 437	C++
 	coreclr.dll!MethodDesc::PrepareCode(PrepareCodeConfig * pConfig) Line 336	C++
 	coreclr.dll!CodeVersionManager::PublishVersionableCodeIfNecessary(MethodDesc * pMethodDesc, CallerGCMode callerGCMode, bool * doBackpatchRef, bool * doFullBackpatchRef) Line 1701	C++
 	coreclr.dll!MethodDesc::DoPrestub(MethodTable * pDispatchingMT, CallerGCMode callerGCMode) Line 2172	C++
 	coreclr.dll!PreStubWorker(TransitionBlock * pTransitionBlock, MethodDesc * pMD) Line 1998	C++
 	coreclr.dll!ThePreStub()	Unknown
 	00007ffdec13f519()	Unknown
...
```

This is due to https://github.com/dotnet/runtime/pull/52427, which introduced this call to compInit():

```
 info.compMethodSuperPMIIndex = g_jitHost->getIntConfigValue(W("SuperPMIMethodContextNumber"), -1);
```

Previously, the get*ConfigValue APIs had only been called by the JIT during jitStartup, which is called only once ever for the load of the JIT. Now, it's being called per compile. This exposes a race condition in the superpmi collection shim jithost implementation, which swaps out the stored MethodContext on a per-compile basis, in the shim `compileMethod` function. But the jithost is shared amongst threads, so one thread can trash the other thread's stashed MethodContext. Previously, it didn't matter because it was never used except for allocateMemory/freeMemory, which doesn't use the stashed MethodContext.

### PR DESCRIPTION
**Pull request by BruceForstall**
PR #52427 introduced a per-compilation call to getIntConfigValue
on "SuperPMIMethodContextNumber". This pseudo-config should not
be collected. It exposed a pre-existing multi-threading issue in
the superpmi collector shim. I got rid of the per-compilation
override of the jithost MethodContext, which is problematic, and
currently unneeded, but documented the considerations around collecting
per-compilation data.

Fixes #53440

### CODE REVIEW

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib PTAL

File: src/coreclr/ToolBox/superpmi/superpmi-shim-collector/jithost.cpp [hunk 1]
```diff
@@ -6,6 +6,20 @@
...
+// There is a single JitHost object created during the one-time initialization of the JIT (function jitStartup),
+// and shared amongst all subsequent compilations. Any calls to the getIntConfigValue/getStringConfigValue
+// APIs get recorded in a single, global MethodContext/CompileResult, and are copied to the
+// per-compilation MethodContext in the shim implementation of compileMethod (using recGlobalContext()).
+// This works because the JIT eagerly asks for all config values once, in the one-time jitStartup
+// function. If the JIT were to ask for config values later, during the per-compilation phase,
+// they would get recorded here in the global MethodContext, and copied to all subsequent
+// compilation MethodContexts. This would be incorrect. A solution would be to use a per-compilation
+// MethodContext in addition to the global MethodContext, but we have to allow for multi-threading. That is,
+// There could be multiple JIT compilations happening concurrently, so we can't just replace the global
```
Comment (kunalspathak) [inline src/coreclr/ToolBox/superpmi/superpmi-shim-collector/jithost.cpp:18]:
```suggestion
// there could be multiple JIT compilations happening concurrently, so we can't just replace the global
```

Comment (kunalspathak) [reply src/coreclr/ToolBox/superpmi/superpmi-shim-collector/jithost.cpp:18]: Do you have a tracking issue to add ability to have config values in per-compilation MethodContext?

**Review by kunalspathak — APPROVED** [review]
LGTM

File: src/coreclr/ToolBox/superpmi/superpmi-shim-collector/jithost.cpp (hunk 1 above)
Comment (BruceForstall) [reply src/coreclr/ToolBox/superpmi/superpmi-shim-collector/jithost.cpp:18]: I'm not sure it's worth having a tracking issue unless we actually need that functionality, which I currently believe we don't.

**Review by echesakov — APPROVED** [review]

**Pull request merged**